### PR TITLE
Delete version checker job if it does not start

### PR DESCRIPTION
**Pull request by davidwding**
Fixes https://github.com/cockroachdb/cockroach-operator/issues/545

The issue is that if a vcheck job is pending for 2 minutes (for whatever reason, like lack of resources), a new vcheck job will be spun up every 2 minutes, leading to numerous pending vcheck jobs.

This PR fixes the immediate issue by deleting the vcheck job if it has not come up successfully.

### CODE REVIEW

File: pkg/resource/cluster.go [hunk 1]
```diff
@@ -154,7 +154,7 @@ func (cluster Cluster) StatefulSetName() string {
 
 func (cluster Cluster) JobName() string {
 	slug.MaxLength = 63
-	return slug.Make(fmt.Sprintf("%s-vcheck-%d", cluster.Name(), getTimeHashInMinutes(time.Now())))
```
[the remark this answers is not in the document]
Comment (chrislovecnm) [reply pkg/resource/cluster.go:157]: I think the best way to figure out if this works is e2e test or a unit test. We need to recreate that it is busted, and then test that we fixed it

**Comment by chrislovecnm** [comment]
Does this fix our problem? Can you rebase please

**Comment by davidwding** [comment]
> Does this fix our problem? Can you rebase please

Sorry about the lack of activity on this PR. Now that I've had some more time to generally read and understand the operator code, I'm confident that this fixes the reported bug and won't cause the potential race condition that you brought up.

**Comment by davidwding** [comment]
Actually, on second thought, this might cause some undesired delays when trying to modify a cluster. Lemme think on it some more and I'll ping back when this is ready for another look.

**Comment by davidwding** [comment]
@chrislovecnm @pseudomuto this is ready for another look. The version checker actor now deletes a job if the pod hasn't come up for 2 minutes. I added a test verifying that when version checker pods are unschedulable, we never end up with more than one.

File: pkg/actor/validate_version.go [hunk 2]
```diff
@@ -139,6 +139,10 @@ func (v *versionChecker) Act(ctx context.Context, cluster *resource.Cluster) err
 		err := WaitUntilJobPodIsRunning(ctx, clientset, job, log)
 		if err != nil {
 			log.Error(err, "job not found")
+			if dErr := deleteJob(ctx, cluster, clientset, job); dErr != nil {
```
Comment (chrislovecnm) [inline pkg/actor/validate_version.go:142]: The log message does not make sense now

File: pkg/actor/validate_version.go [hunk 3]
```diff
@@ -293,15 +300,9 @@ func (v *versionChecker) Act(ctx context.Context, cluster *resource.Cluster) err
...
-	//delete the job only if we have managed to get the version and we do not have any errors
-	err = clientset.BatchV1().Jobs(cluster.Namespace()).Delete(ctx, job.Name, metav1.DeleteOptions{
-		GracePeriodSeconds: ptr.Int64(5),
-		PropagationPolicy:  &dp,
-	})
-	if err != nil {
-		log.Error(err, "failed to delete the job")
+	// If we got here, the version checker job was successful. Delete it.
+	if dErr := deleteJob(ctx, cluster, clientset, job); dErr != nil {
+		log.Error(dErr, "failed to delete the job")
```
Comment (chrislovecnm) [inline pkg/actor/validate_version.go:305]: Does log.Error return an error? Wondering if we should return here

**Review by chrislovecnm — COMMENTED** [review]
Just a nitpick on a log message. Otherwise looks great

File: pkg/actor/validate_version.go (hunk 3 above)
Comment (davidwding) [reply pkg/actor/validate_version.go:305]: log.Error doesn't return an error, but do you think returning here is the right thing to do? I could see an argument either way - let the version checker finish and potentially have a dangling job, or just return and say the version checker failed.

File: pkg/actor/validate_version.go (hunk 2 above)
Comment (davidwding) [reply pkg/actor/validate_version.go:142]: updating the error on the previous line to be more accurate

File: pkg/actor/validate_version.go (hunk 3 above)
Comment (udnay-crl) [reply pkg/actor/validate_version.go:305]: If we error here I would worry that we can't move forward with the next step. Having a dangling job is probably ok IMO, better than having version checker come back as ok but fail on clean up. Maybe add more context in the error message. Check passed but delete failed.

Comment (davidwding) [reply pkg/actor/validate_version.go]: I've updated the error message as we just discussed. This PR should be good to go.

**Review by chrislovecnm — APPROVED** [review]
Lgtm

**Pull request merged**